spec/notification_center_spec.rb

```diff
@@ -0,0 +1,484 @@
+         @callback_reference
+        )).to eq(2)
+  
+        # add another callback for NOTIFICATION_TYPES::ACTIVATE
+        expect(@inner_notification_center.add_notification_listener(
+         Optimizely::NotificationCenter::NOTIFICATION_TYPES[:ACTIVATE],
+         @callback_reference_second
+        )).to eq(3)
+        # Verify that notifications length for NOTIFICATION_TYPES::ACTIVATE is 2
+        expect(@inner_notification_center.notifications[
+         Optimizely::NotificationCenter::NOTIFICATION_TYPES[:ACTIVATE]].length).to eq(2)
+        # Verify that notifications length for NotificationType::TRACK is 1
+        expect(@inner_notification_center.notifications[
+         Optimizely::NotificationCenter::NOTIFICATION_TYPES[:TRACK]].length).to eq(1)
+      end
+      
+      it 'should not remove callback for empty notification ID' do
+        expect(@inner_notification_center.remove_notification_listener(nil)).to eq(nil)
+        expect(spy_logger).to have_received(:log).once
+                               .with(Logger::ERROR, 'Notification ID can not be empty.')
+
+        expect(@inner_notification_center.notifications[Optimizely::NotificationCenter::NOTIFICATION_TYPES[:ACTIVATE]].length).to eq(2)
+        expect(@inner_notification_center.notifications[Optimizely::NotificationCenter::NOTIFICATION_TYPES[:TRACK]].length).to eq(1)
+      end
+      
+      it 'should not remove callback for an invalid notification ID' do
+        expect(@inner_notification_center.remove_notification_listener(4))
+         .to eq(false)
+        expect(@inner_notification_center.notifications[Optimizely::NotificationCenter::NOTIFICATION_TYPES[:ACTIVATE]].length).to eq(2)
+        expect(@inner_notification_center.notifications[Optimizely::NotificationCenter::NOTIFICATION_TYPES[:TRACK]].length).to eq(1)
+      end
+
+      it 'should remove callback for a valid notification ID' do
+        expect(@inner_notification_center.remove_notification_listener(3))
+         .to eq(true)
+        expect(@inner_notification_center.notifications[Optimizely::NotificationCenter::NOTIFICATION_TYPES[:ACTIVATE]].length).to eq(1)
+        expect(@inner_notification_center.notifications[Optimizely::NotificationCenter::NOTIFICATION_TYPES[:TRACK]].length).to eq(1)
+      end
+
+      it 'should not remove callback once a callback has already been removed against a notification ID' do
+        expect(@inner_notification_center.remove_notification_listener(3))
+         .to eq(true)
+        expect(@inner_notification_center.remove_notification_listener(3))
+         .to eq(false)
+        expect(@inner_notification_center.notifications[Optimizely::NotificationCenter::NOTIFICATION_TYPES[:ACTIVATE]].length).to eq(1)
+        expect(@inner_notification_center.notifications[Optimizely::NotificationCenter::NOTIFICATION_TYPES[:TRACK]].length).to eq(1)
+      end
+      
+      it 'should not remove notifications for an invalid notification type' do
```

Comment:
Duplicate test case? See Line 222 it 'should not remove callback for an invalid notification ID' do